Fix #21623 [BUG]: Discrepancy Between Topic Preview and Learner View

mohitagarwal24 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #21623 .
2. This PR does the following: 
-  Completely remove the usage of `previewMode`.
-  Reintroduce the original logic from before #15192, with additional functionality to ensure that `startButtonIsDisabled` is set to `false` when we successfully verify that `questionsAreAvailable === true`.
-  Adding logic to get the `topicUrlFragment` and `classroomUrlFragement` even when the page url starts with `\topic_editor`.
-  Eliminate the condition involving `this.previewMode` in branching logic and simplify it to:
   ```typescript
   this.displayArea === 'topicViewer' && this.urlService.getPathname().startsWith('/learn')
   ```

This approach ensures that:
- The `topicUrlFragment` and `classroomUrlFragment` values are correctly populated in both cases when url starts with `/learn` and `/topic_editor`.
- Errors for URLs starting with `/topic_editor` are avoided.
- The `startButtonIsDisabled` state is appropriately managed based on the availability of questions.

3. (For bug-fixing PRs only) The original bug occurred because: 

The cause of bug was that the `previewMode` variable was causing the problem. It was hardcoded as true in the `core/templates/pages/topic-editor-page/preview-tab/topic-preview-tab.component.ts` file. This was affecting the logic for enabling the "Start" button in `/home/mohitagarwal24/opensource/oppia/core/templates/pages/topic-viewer-page/practice-tab/practice-tab.component.html`. The button depends on two boolean values: `isStartButtonDisabled()` and `questionsStatusCallIsComplete`. One of them should be false for the button to be enabled. Initially, both are `true`, but after selecting a topic, `questionsStatusCallIsComplete` should become `true`, and `isStartButtonDisabled()` should become `false`.
Moreover, even when this was solved, when clicking on `start button`, it would not lead us to the practice session because there is no logic defined for population of url with correct `topicUrlFragment` and `classroomUrlFragment` when pathname starts with `\topic-editor`. This PR fixes that issue as well. For more details please view discussion on #21623.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).


## Proof that changes are correct

https://github.com/user-attachments/assets/141c6ebf-e0b9-4711-b996-11744cda83ce

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->


## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

mohitagarwal24 committed:
fixing start button disabled and url not populating properly for topic-editor page

mohitagarwal24 committed:
fix type check error

mohitagarwal24 committed:
fix frontend testing

mohitagarwal24 committed:
More tests added to increase coverage

mohitagarwal24 committed:
Linter checks corrected

seanlip commented:
Assigning @amyyeung17 for first pass (shadow reviewer).

amyyeung17 commented on core/templates/pages/topic-editor-page/preview-tab/topic-preview-tab.component.html:
```diff
@@ -59,7 +59,7 @@
       <practice-tab [subtopicsList]="subtopics"
                     [topicName]="topicName"
-                    [previewMode]="true">
+                    [startButtonIsDisabled]="true">
```
Nice improvement with the variable name. :)

Does startButtonIsDisabled = true represent the default case when loading the practice tab? If the practice tab always loads with no subtopics selected, could we initialize it directly in the component instead (at line 52 in the .ts file)?

amyyeung17 replied:
Also, small reminder to make sure to double check that changes here should be reflected for all practice-tab in the codebase.

mohitagarwal24 replied:
Hi @amyyeung17 I confirmed the behaviour with `/learn/math/dummy-topic-one/practice` page and it is same as you mentioned so we can also initialise `startButtonIsDisabled` to true directly in .ts file. As I am new to this org I am not very much aware of coding styles so can you please confirm whether I do this or not.

amyyeung17 replied:
No worries @mohitagarwal24. Developing is all about learning. :) These are just general best practices, which you'll learn as you go.

mohitagarwal24 replied:
Thanks @amyyeung17. Now if everything looks good, can I push these changes ?

amyyeung17 commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.component.ts:
```diff
@@ -172,6 +191,9 @@ export class PracticeTabComponent implements OnInit, OnDestroy {
         .then(questionCount => {
           this.questionsAreAvailable = questionCount > 0;
           this.questionsStatusCallIsComplete = true;
+          if (this.questionsAreAvailable) {
+            this.startButtonIsDisabled = false;
+          }
```
So after clicking a subtopic, checkIfQuestionExists is called. If there are questions available (questionsAreAvailable = true), we set startButtonIsDisabled = false. Since startButtonIsDisabled has our desired state, could we return that instead of running an extra for loop at lines 172-176 in the function isStartButtonDisabled? Alternatively, do we even need this function at all?

amyyeung17 replied:
If a topic is unchecked or there are no available questions, should the start button also be disabled? Are these correctly reflected in the function?

mohitagarwal24 replied:
We need isStartButtonDisabled() because we need to check whether topics are selected or not which is maintained by `selectedtopicIndices`. If we remove that function or loop then once we select the topic the start button gets enabled but when we deselect it, start button does not get disabled. 


https://github.com/user-attachments/assets/1a7185a7-f92a-449e-aac2-8ab52943ead2

amyyeung17 replied:
I suggest taking a look at line 28 in the practice-tab.html. checkIfQuestionExists actually passes selectedSubtopicIndices as its argument, so subtopicIndices = selectedSubtopicIndices. It appears we're doing the check in the function already, so setting startButtonIsDisabled = false in the else statement will handle this case. :)

mohitagarwal24 replied:
```typescript
if (skillIds.length > 0) {
      this.questionBackendApiService
        .fetchTotalQuestionCountForSkillIdsAsync(skillIds)
        .then(questionCount => {
          this.questionsAreAvailable = questionCount > 0;
          this.questionsStatusCallIsComplete = true;
          if (this.questionsAreAvailable) {
            this.startButtonIsDisabled = false;
          }
          else {
            this.startButtonIsDisabled = true;
          }
        });
    } else {
      this.startButtonIsDisabled = true;
      this.questionsAreAvailable = false;
      this.questionsStatusCallIsComplete = true;
    }
```

Hi @amyyeung17, I did this change and completely removed isStartButtonDisabled(). Now does it looks good?

amyyeung17 replied:
I thought I mentioned this but I guess I forgot. You can rewrite this.startButtonIsDisabled = !this.questionsAreAvailable instead of if statement. It's clear enough based on the variable names.

mohitagarwal24 replied:
Done.

amyyeung17 commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.component.ts:
```diff
@@ -101,11 +102,29 @@ export class PracticeTabComponent implements OnInit, OnDestroy {
+        this.editableTopicBackendApiService.fetchTopicAsync(this.topicId);
+      topicDataPromise
+        .then(topicData => {
+          this.topicUrlFragment = topicData.topicDict.url_fragment;
```
topicUrlFragment is a input. Best to avoid directly modifying input variables, which can cause side effects. If topicUrlFragment isn't being passed, we can convert this to a local variable. This applies to other input variables like classroomUrlFragment as well.

mohitagarwal24 replied:
Yes actually topicUrlFragment and classroomUrlFragment are populated in ngOnInit only, their values anyway does not depend on what they got from parent components so this can be done. Thanks!

amyyeung17 replied:
@mohitagarwal24 Hi, please don't resolve the conversations (rule 5: https://github.com/oppia/oppia/wiki/Rules-for-making-PRs). Please also reassign me when you're requesting a review, thanks!

mohitagarwal24 replied:
Hi @amyyeung17 Actually we have to keep topicUrlFragment and classroomUrlFragment as it is being passed in `old-progress-tab.component.html`. But changing these variables is dependent on displayArea and we are only changing values of these variables when displayArea==='topicViewer'. All other changes I have made and pushed.

amyyeung17 replied:
@seanlip Should we leave a comment or file a issue for theses changes after the learner dashboard migration? We shouldn't be modifying input values, so it makes sense to convert them to component variables later down the road.

mohitagarwal24 replied:
@amyyeung17 Can we introduce two new variables like `_topicUrlFragment` and `_classroomUrlFragment` and on initialisation assign input variable values to these local variables and then use these local variables throughout the file ??

amyyeung17 replied:
@mohitagarwal24 That is true. I would say go ahead and do it. We can file an issue to eliminate the input variables for the future. @seanlip — is that alright?

seanlip replied:
Thanks for looping me in, @amyyeung17!

Actually, I don't actually understand what is going on here. Either we are getting the topic URL fragment from the backend call, or we are getting the topic URL fragment from the input. Why are we trying to do both? Shouldn't we be doing one or the other? There should be a single source of truth.

@amyyeung17 re your question, in general, I'd say that this PR should properly fix the issue, and after it's merged we ideally don't need to revisit this part of the code. So I would suggest that the problem noted here gets resolved within the context of this PR. (In general, let's not create new issues unless it's something that's really out of scope, which I don't think this qualifies as because it's code that's added in this PR -- this guidance applies to other PRs as well.) 

Also, re your earlier comment in this thread, I don't think this has anything to do with the learner dashboard migration?

amyyeung17 commented:
> Assigning @amyyeung17 for first pass (shadow reviewer).

Done!

seanlip commented:
@mohitagarwal24 Please read step 5 of https://github.com/oppia/oppia/wiki/Rules-for-making-PRs#step-5-address-review-comments-until-all-reviewers-approve more carefully. Push your changes, address all comments, and assign the reviewer back for a review. You don't need to ask for permission to push. You should also ensure that whoever's supposed to take the next step on the PR is assigned to it, otherwise they might miss it. (We ignore the "Reviewers" pane on GitHub.) All of this is explained in the wiki page above.

Note that, since @amyyeung17 is shadow-reviewing, she will assign this to me once she is happy with your final code.

Thanks.

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page
merge with upstream-develop

mohitagarwal24 committed:
address review comments

mohitagarwal24 commented:
@amyyeung17 PTAL! Thanks.

amyyeung17 commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.component.ts:
```diff
@@ -193,9 +181,12 @@ export class PracticeTabComponent implements OnInit, OnDestroy {
             this.startButtonIsDisabled = false;
+          } else {
+            this.startButtonIsDisabled = true;
           }
```
Instead of setting this to true/false, we can do this.startButtonIsDisabled = !this.questionsAreAvailable. (so if questions are available, it is false that the start button is disabled).

amyyeung17 commented:
@seanlip The old code had those variables as input values and would modify them. Only the practice-tab in the old-progress-tab of the learner dashboard actually passed anything. I'm not exactly sure why, but it doesn't matter because the old-progress-tab is obsolete. However, removing the inputs completely might break the current dashboard.  We can check to see if they're necessary

seanlip commented:
> We can check to see if they're necessary first. If they are, we can either change the old-progress-tab's code to break the dependence or wait until the new dashboard.

Yup, I think that sounds like a good way to proceed. I suggest that we try to understand, first of all, whether we can do all of this without any backend calls, both in the new + old dashboard, and modify the code for both as needed. If that is not possible then we can consider approaches for doing a backend call such that there is still a clear single source of truth in all cases.

mohitagarwal24 committed:
address review comments

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

mohitagarwal24 committed:
linter checks corrected

mohitagarwal24 committed:
strict type checks

mohitagarwal24 committed:
fix frontend tests

mohitagarwal24 commented:
Hey @seanlip @amyyeung17 PTAL! I have removed the backend calls which were changing the values of input variables and modified the parent files to correctly bind the input variables to pass the correct data and maintain a single source of truth. Let me know if it looks good.

amyyeung17 commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.component.ts:
```diff
@@ -101,12 +98,6 @@ export class PracticeTabComponent implements OnInit, OnDestroy {
       false
     );
     this.clientWidth = window.innerWidth;
-    if (this.displayArea === 'topicViewer' && !this.previewMode) {
-      this.topicUrlFragment =
-        this.urlService.getTopicUrlFragmentFromLearnerUrl();
-      this.classroomUrlFragment =
-        this.urlService.getClassroomUrlFragmentFromLearnerUrl();
-    }
```
I think we might have a small misunderstanding. The original problem was whether having input variables were necessary at all since this part here sets it up again. In other words, we need to check if these functions can replace passing topicUrlFragment and classroomUrlFragment through practice-tab in the old-progress-tab. If not, we can remove the remove the input variables and just do the functions call.

I suggest checking if these calls are retrieving the same values we are passing through the old practice tab first. If so, then the inputs aren't necessary. I'm not 100% sure what this.displayArea and this.previewMode refers to, but make sure this doesn't interfere with how the variables are being set for the practice-tab in the old-progress-tab. There might be a reason why the previous code chose to pass the variables and not for the practice-tab used elsewhere.

mohitagarwal24 replied:
The methods which were earlier defined (getTopicUrlFragmentFromLearnerUrl() and getClassroomUrlFragmentFromLearnerUrl()) were solely focused to populate data for learner view not the editor view and same functions were used to get data for topic editor view as well, which was inappropriate. But now the data gets passed directly through topic-editor as well as topic-viewer depending on what input variables they pass.

Also regarding this.displayArea and this.previewMode, I am not sure why they were introduced (maybe to reduce unneccessary calls) because in the original code if we directly populated this.topicUrlFragment and this.classroomUrlFragment then also it would behave in same way as it did with if statement.

mohitagarwal24 replied:
Hi @amyyeung17 @seanlip PTAL. Thanks!

amyyeung17 replied:
So just to double-check, the topic.getUrlFragment() and topicEditorStateService.getClassroomUrlFragment() in topic-viewer returns something different when compared to getClassroomUrlFragmentFromLearnerUrl() and getTopicUrlFragmentFromLearnerUrl() in practice-tab?

mohitagarwal24 replied:
@amyyeung17 
When user is in topic editor view the url starts with `/topic_editor` but in learner view url starts with `/learn` and practice-tab is there in both cases and its behaviour should be same in both the cases. But the functions getClassroomUrlFragmentFromLearnerUrl() and getTopicUrlFragmentFromLearnerUrl() only works with `/learn`  and `/explore` url not with `/topic_editor`.  

Also, regarding your question topic.getUrlFragment() and topicEditorStateService.getClassroomUrlFragment() returns the same thing as getTopicUrlFragmentFromLearnerUrl() and getClassroomUrlFragmentFromLearnerUrl(). It's just that the latter methods use url params but the former ones don't. Also this way input variables will not get modified in practice-tab.component.ts

amyyeung17 replied:
Hm okay. It seems like we need to keep the input variables for the old-progress-tab. It'll be obsolete after the new learner dashboard is released, so we won't need them later on. For now, this is perfectly fine. Good job.

amyyeung17 replied:
Wait, sorry disregard the previous comment and new review. Did you check if  urlService.getTopicUrlFragmentFromLearnerUrl() works with the topic? There's an if statement that recognizes 'topic_url_fragment'.

mohitagarwal24 replied:
Sorry I think I am unable to understand you question. Can you specify the line of if statement you are referring to?

amyyeung17 replied:
Oh, I finally understand where you're coming from. I thought urlServices had a function that supported retrieving the classroom for topic editors. Good job on identifying the differences. Can you please remove displayArea too since we don't need it anymore? I believe this is the last thing I have to comment on.

mohitagarwal24 replied:
Done.

mohitagarwal24 replied:
Hi @amyyeung17 Although displayArea is not directly used in `practice-tab.component.ts` but it is still binded in ngIf* in `practice-tab.component.html` which decided what UI to render so I don't think removing that would be a good thing to do?

amyyeung17 replied:
Hi @mohitagarwal24, nice catch. Hm, it seems like displayArea was only used to differentiate old-progress-tab. We can actually do the check within the practice-tab and remove the dependencies, but I think it might easier just to pass the argument through displayArea. Can you undo erasing displayArea='progressTab' in old-progress-tab?

Vir-8 replied:
Hi, can you confirm if this comment has been resolved and we're good to go? Thanks! @mohitagarwal24 @amyyeung17

mohitagarwal24 replied:
@Vir-8 yes it was done. Please see https://github.com/oppia/oppia/pull/21874/commits/ce539b1fcc726ad5973398aeb558ab0bef4539df

amyyeung17 commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.component.ts:
```diff
@@ -49,7 +47,7 @@ export class PracticeTabComponent implements OnInit, OnDestroy {
   @Input() topicName!: string;
   @Input() subtopicsList!: Subtopic[];
-  @Input() previewMode: boolean = false;
+  @Input() startButtonIsDisabled: boolean = true;
```
startButtonIsDisabled isn't an argument anymore, so we should change this to a component variable. Inputs shouldn't be modified.

mohitagarwal24 replied:
Done.

mohitagarwal24 committed:
remove unnecessary input variables

mohitagarwal24 commented:
@seanlip @amyyeung17 PTAL! Thanks.

amyyeung17 commented on core/templates/pages/topic-editor-page/preview-tab/topic-preview-tab.component.html:
```diff
@@ -59,7 +59,8 @@
       <practice-tab [subtopicsList]="subtopics"
                     [topicName]="topicName"
-                    [previewMode]="true">
+                    [topicUrlFragment]="topicUrlFragment"
+                    [classroomUrlFragment]="classroomUrlFragment">
```
Why are we passing topicUrlFragment and classroomUrlFragment? Was it not working before without it?

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

amyyeung17 commented:
Hi @mohitagarwal24, please take a look at https://github.com/oppia/oppia/pull/21874#discussion_r1975858163 and https://github.com/oppia/oppia/pull/21874#discussion_r1975861627. These should be the last comments I have :)

mohitagarwal24 committed:
address review comments

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

mohitagarwal24 committed:
frontend checks

mohitagarwal24 committed:
correction in displayArea

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

mohitagarwal24 commented:
Hi @amyyeung17 PTAL! https://github.com/oppia/oppia/pull/21874#discussion_r1976333658

seanlip commented:
Just a note, I don't think this is quite "done" yet -- @amyyeung17 please feel free to assign me when it is.

(Also just a note, if you are waiting for @mohitagarwal24 to respond then please ensure he is assigned to the issue and you're not -- currently, it says you are the assignee. Thanks!)

mohitagarwal24 committed:
redo changes in displayArea in old-progess-tab

mohitagarwal24 committed:
oMerge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

amyyeung17 commented:
@seanlip Okay, now it should be LGTM. Thanks!

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

seanlip commented:
Sorry, don't have enough time to look at this today. Will try to do so tomorrow.

seanlip reviewed (approved):
In my opinion, this looks excellent -- great description, and it's nice to see the removal of the branching.

Well done, and thanks, @mohitagarwal24 and @amyyeung17! LGTM.

Vir-8 commented:
Hi @mohitagarwal24, PTAL https://github.com/oppia/oppia/pull/21874#discussion_r1986212964. Thanks!

mohitagarwal24 commented:
Hi @Vir-8 PTAL! Thanks https://github.com/oppia/oppia/pull/21874#discussion_r1986217519

Vir-8 reviewed (approved):
LGTM!

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

mohitagarwal24 committed:
Merge remote-tracking branch 'upstream/develop' into fix-discrepancy-between-topic-editor-and-learner-page

seanlip commented:
Just a reminder to please follow the instructions [here](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR) if your CI checks are failing (whether in the regular view or in the merge queue). Thanks!

HardikGoyal2003 commented:
@mohitagarwal24 Just a reminder to please follow the instructions [here](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR) if your CI checks are failing (whether in the regular view or in the merge queue).

If this is not done within 2 days, I will close this PR. Thanks!

mohitagarwal24 commented:
> @mohitagarwal24 Just a reminder to please follow the instructions [here](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR) if your CI checks are failing (whether in the regular view or in the merge queue).
> 
> If this is not done within 2 days, I will close this PR. Thanks!

Hi, actually on merge queue every time a different test is failing like when you first added to merge-queue "
Acceptance (logged-out-user/play-through-lesson-while-getting-feedback-and-hints)" then another time "E2E(publication)" failed then "Lighthouse" failed then another time "frontend unit test failed", (while these all passed here) so that's why I am unable to narrow down what exactly is the issue. Can you suggest some way to figure it out why this is happening?

seanlip commented:
> > > @mohitagarwal24 Just a reminder to please follow the instructions [here](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR) if your CI checks are failing (whether in the regular view or in the merge queue).
> > > If this is not done within 2 days, I will close this PR. Thanks!
> > 
> > 
> > Hi, actually on merge queue every time a different test is failing like when you first added to merge-queue " Acceptance (logged-out-user/play-through-lesson-while-getting-feedback-and-hints)" then another time "E2E(publication)" failed then "Lighthouse" failed then another time "frontend unit test failed", (while these all passed here) so that's why I am unable to narrow down what exactly is the issue. Can you suggest some way to figure it out why this is happening?
> 
> Hi @seanlip @HardikGoyal2003 Can you please take a look at this comment.

Have you read https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR ? We updated it recently; see the last question at the bottom of that page.

mohitagarwal24 commented:
> > > > @mohitagarwal24 Just a reminder to please follow the instructions [here](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR) if your CI checks are failing (whether in the regular view or in the merge queue).
> > > > If this is not done within 2 days, I will close this PR. Thanks!
> > > 
> > > 
> > > Hi, actually on merge queue every time a different test is failing like when you first added to merge-queue " Acceptance (logged-out-user/play-through-lesson-while-getting-feedback-and-hints)" then another time "E2E(publication)" failed then "Lighthouse" failed then another time "frontend unit test failed", (while these all passed here) so that's why I am unable to narrow down what exactly is the issue. Can you suggest some way to figure it out why this is happening?
> > 
> > 
> > Hi @seanlip @HardikGoyal2003 Can you please take a look at this comment.
> 
> Have you read https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR ? We updated it recently; see the last question at the bottom of that page.

Yes I have gone through that page but what my concern is that why every time a different test is failing (that might be a flake) but it has occurred many times so there might be something else which is incorrect.

mohitagarwal24 commented:
Ok thanks I will try to resolve it ASAP.
